kairon/shared/utils.py

```diff
@@ -1757,25 +1757,52 @@ def get_graph(steps: List) -> DiGraph:
         vertices = {}
         for story_step in steps:
             vertices[story_step['step']['node_id']] = KaironStoryStep(story_step["step"]["name"],
-                                                                   story_step["step"]["type"])
+                                                                    story_step["step"]["type"], story_step['step']['node_id'],
+                                                                    story_step["step"].get('value'))
         for story_step in steps:
             story_step_object = vertices[story_step["step"]["node_id"]]
             for connected_story_step in story_step['connections'] or []:
                 if connected_story_step['node_id'] in vertices.keys():
                     connection_object = vertices[connected_story_step["node_id"]]
                 else:
-                    connection_object = KaironStoryStep(connected_story_step["name"], connected_story_step["type"])
+                    connection_object = KaironStoryStep(connected_story_step["name"], connected_story_step["type"],
+                                                        connected_story_step["node_id"])
                 graph.add_edge(story_step_object, connection_object)
         return graph
 
     @staticmethod
-    def validate_steps(steps: List):
+    def get_leaf_nodes(story_graph: DiGraph):
+        return [node for node in story_graph.nodes() if story_graph.out_degree(node) == 0]
+
+    @staticmethod
+    def get_leaf_nodes_with_intent(story_graph: DiGraph):
+        return [node for node in story_graph.nodes() if story_graph.out_degree(node) == 0
+                                    and node.step_type == 'INTENT']
+
+    @staticmethod
+    def get_source_node(story_graph: DiGraph):
+        return [x for x in story_graph.nodes() if story_graph.in_degree(x) == 0]
+
+    @staticmethod
+    def validate_graph_slot(story_graph: DiGraph):
+        for node in story_graph.nodes():
```

Comment:
not worth iterating again just to validate slot events. try using the loop in validate steps line 1819.